Default gateway.auto_import_dangling_indices to false

pugnascotia opened the pull request:
Backport of #58898.

Part of #48366. Now that there is a dedicated API for dangling indices, the auto-import behaviour can default to off. Also add a note to the breaking changes for 7.9.0.

pugnascotia committed:
Default gateway.auto_import_dangling_indices to false (#58898)

Part of #48366. Now that there is a dedicated API for dangling indices,
the auto-import behaviour can default to off.

pugnascotia committed:
Add a breaking change note about dangling indices

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Distributed)

pugnascotia commented:
@elasticmachine run elasticsearch-ci/oss-distro-docs

lockewritesdocs commented on docs/reference/release-notes/7.9.asciidoc:
```diff
@@ -22,3 +22,13 @@ Snapshot restore throttling::
   that no behavioral change will be observed by clusters where the recovery
   and restore settings had not been adapted from the defaults. {es-pull}58658[#58658]
+
+Dangling indices::
+* A new <<dangling-indices-api,Dangling indices API>> is now available. You
+  can use it to list, import or delete
+  <<dangling-indices,dangling indices>>. Because of this, automatically
+  importing dangling indices is now disabled by default. There are several
+  issues with this behavior, and you should use the dedicated API instead.
+  If you need to re-enable auto-imports, set
+  `gateway.auto_import_dangling_indices` to `true`.  Auto-imports are now
+  deprecated and will be removed in 8.0 {es-pull}58176[#58176] {es-pull}58898[#58898] (issue: {es-issue}48366[#48366])
```
```suggestion
Dangling indices::
* A new <<dangling-indices-api,Dangling indices API>> is now available, which you
  can use to list, import, or delete
  <<dangling-indices,dangling indices>>. Because importing dangling indices
  into the cluster using `gateway.auto_import_dangling_indices` is unsafe, use the dedicated API instead.
  
  Automatically importing dangling indices is now deprecated, disabled by default, and will be removed in (es) 8.0. If you need to re-enable auto-imports, set
  `gateway.auto_import_dangling_indices` to `true`
   {es-pull}58176[#58176] {es-pull}58898[#58898] (issue: {es-issue}48366[#48366])
```

lockewritesdocs replied:
>Because of this, automatically
  importing dangling indices is now disabled by default.

I think that we should clarify this language to explain what "this" is. I reused some of the language from [58751](https://github.com/elastic/elasticsearch/pull/58751/files#diff-d5961dc535d48b0d23e553177bc81b2fR61-R62), which emphasizes that using `gateway.auto_import_dangling_indices` is unsafe. Please let me know if I misconstrued the original meaning of this explanation.

lockewritesdocs reviewed (approved):
Added a suggested revision to the release notes, but LGTM otherwise. Thanks!

pugnascotia committed:
Update docs/reference/release-notes/7.9.asciidoc

Co-authored-by: Adam Locke <adam.locke@elastic.co>

DaveCTurner reviewed (changes requested):
Since this is a breaking change I think we should document it in `docs/reference/migration/migrate_7_9.asciidoc` rather than `docs/reference/release-notes/7.9.asciidoc`. I suppose we can do both, but that seems excessive to me.

lockewritesdocs commented:
> Since this is a breaking change I think we should document it in `docs/reference/migration/migrate_7_9.asciidoc` rather than `docs/reference/release-notes/7.9.asciidoc`. I suppose we can do both, but that seems excessive to me.

That's a good call @DaveCTurner 👍 Thanks for pointing out that change.

pugnascotia committed:
Update 7.9 migration notes

Move most of the breaking changes regarding dangling indices in the 7.9
release notes to the 7.9 migration doc. Abbreviate the release ntoes
content and point to the migration notes instead.

pugnascotia commented:
Thanks @DaveCTurner. I didn't feel entirely comfortable with completely removing the content from the release notes, so I abbreviated it and pointed it to the migration notes.

DaveCTurner commented on docs/reference/migration/migrate_7_9.asciidoc:
```diff
@@ -67,5 +68,28 @@ setting. You may use `script.context.$CONTEXT.max_compilations_rate` for the par
+[%collapsible]
+====
+*Details* +
+<<dangling-indices,Dangling indices>> will no longer be imported
```
I think it'd help to move some things around here: IMO the `Details` section describes the change and the `Impact` tells the user what to do about it. How about this:

```diff
diff --git a/docs/reference/migration/migrate_7_9.asciidoc b/docs/reference/migration/migrate_7_9.asciidoc
index e60c4f3588d..acfd11fc38b 100644
--- a/docs/reference/migration/migrate_7_9.asciidoc
+++ b/docs/reference/migration/migrate_7_9.asciidoc
@@ -80,16 +80,18 @@ context.  For example, for the `processor_conditional` context, use
 [%collapsible]
 ====
 *Details* +
-<<dangling-indices,Dangling indices>> will no longer be imported
-automatically by default. A new
-<<dangling-indices-api,Dangling indices API>> is now available, which you
-can use to list, import, or delete. Because importing dangling indices
-automatically into the cluster is unsafe, use the dedicated API instead.
+Automatically importing <<dangling-indices,dangling indices>> into the cluster
+is unsafe and is now disabled by default. This feature will be removed entirely
+in {es} 8.0.0.

 *Impact* +
-Automatically importing dangling indices is now deprecated, disabled by
-default, and will be removed in (es) 8.0. If you need to re-enable
-auto-imports, set `gateway.auto_import_dangling_indices` to `true`.
+Use the <<dangling-indices-api,Dangling indices API>> to list, delete or import
+any dangling indices manually.
++
+Alternatively you can enable automatic imports of dangling indices, recovering
+the unsafe behaviour of earlier versions, by setting
+`gateway.auto_import_dangling_indices` to `true`. This setting is deprecated
+and will be removed in {es} 8.0.0. We do not recommend using this setting.

 ====
 //end::notable-breaking-changes[]
```

DaveCTurner commented on docs/reference/release-notes/7.9.asciidoc:
```diff
@@ -22,3 +22,9 @@ Snapshot restore throttling::
+
+Dangling indices::
+* Automatically importing dangling indices is now deprecated, disabled by
+  default, and will be removed in (es) 8.0. See the
```
I think you meant this:

```suggestion
  default, and will be removed in {es} 8.0. See the
```

DaveCTurner reviewed (commented):
Left some suggestions

pugnascotia committed:
Apply further suggestions

DaveCTurner reviewed (approved):
LGTM

pugnascotia committed:
fix typo

pugnascotia commented:
Cherry-picked to `7.9` in 84f4635f30a
